### PR TITLE
DIV-4583 - fix incorrect step skipping on amend petition

### PR DESCRIPTION
**Pull request by hosainnet**
# Description

This is to fix a scenario where skipping answered questions still stops
at Check Your Answers, because the user could have answered a question,
then clicked saved an resume, which means they come back, the page is
skipped and the case is not submitted

Fixes #DIV-4583

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -76,7 +76,7 @@ module.exports = function(grunt) {
         options: {
           nodeArgs: ['--trace-warnings', '--inspect'],
           ext: 'js, json, yaml',
-          ignore: ['node_modules/**', 'app/assets/**', 'public/**'],
+          ignore: ['node_modules/**', 'app/assets/**', 'public/**', 'functional-output/**', 'coverage/**', 'test/**'],
```
Comment (hosainnet) [inline Gruntfile.js]: This fixes a lot of flakiness locally because those paths constantly change, causing the app to restart and making the tests fail

File: app/core/helpers/steps.js [hunk 2]
```diff
@@ -27,6 +27,8 @@ const findNextUnAnsweredStep = function* (step, session = {}) {
 
   if (!nextStep) {
     return step;
+  } else if (!nextStep.isSkipWhenValid()) {
+    return nextStep;
```
Comment (hosainnet) [inline app/core/helpers/steps.js]: We don't continue finding steps if we arrive at a step which shouldn't be skipped

File: package.json [hunk 3]
```diff
@@ -25,7 +25,7 @@
     "test:smoke": "./bin/run-smoke-tests.sh",
     "test:functional": "./bin/run-functional-tests.sh",
     "test:crossbrowser": "./bin/run-crossbrowser-tests.sh",
-    "test-unit": "NODE_ENV=testing LOG_LEVEL=OFF NODE_PATH=. mocha './app/**/*.test.js' --reporter spec --recursive -t 3000",
+    "test-unit": "NODE_ENV=testing LOG_LEVEL=OFF NODE_PATH=. mocha './app/**/*.test.js' --reporter spec --recursive -t 10000",
```
Comment (hosainnet) [inline package.json]: I found this helpped a little locally

File: test/end-to-end/pages/index/index.js [hunk 4]
```diff
@@ -16,22 +16,13 @@ function startApplication(ignoreIdamToggle = false) {
...
+async function startApplicationWith(sessionName) {
 
   let I = this;
 
   I.amOnLoadedPage('/index');
   I.startApplication();
-  I.haveABasicSession('basicDivorceSessionData');
-}
-
-async function startApplicationWithAnAmendPetitionSession() {
```
Comment (hosainnet) [inline test/end-to-end/pages/index/index.js]: Refactored these to be more generic

File: app/steps/check-your-answers/index.js [hunk 5]
```diff
@@ -304,4 +304,8 @@ module.exports = class CheckYourAnswers extends ValidationStep {
         res.redirect('/generic-error');
       });
   }
+
+  isSkipWhenValid() {
```
Comment (dtaylor7) [inline app/steps/check-your-answers/index.js]: would never skip cya anyway, can remove this

File: app/core/helpers/steps.js [hunk 6]
```diff
@@ -27,6 +27,8 @@ const findNextUnAnsweredStep = function* (step, session = {}) {
 
   if (!nextStep) {
     return step;
+  } else if (!nextStep.isSkipWhenValid()) {
```
Comment (tchow8) [inline app/core/helpers/steps.js]: Maybe just me, but isSkippableWhenValid or shouldSkipWhenValid sounds better to me? haha but that is preference.

Comment (hosainnet) [reply app/core/helpers/steps.js]: Yep agreed, will change it 👍

File: test/end-to-end/pages/index/index.js (hunk 4 above)
Comment (vianvi) [reply test/end-to-end/pages/index/index.js]: can we run all tests on PR?

Comment (hosainnet) [reply test/end-to-end/pages/index/index.js]: Sure, just found a small issue with a unit test, I'll enable all tests on PR

File: app/steps/check-your-answers/index.js (hunk 5 above)
Comment (hosainnet) [reply app/steps/check-your-answers/index.js]: So the scenario is, if you tick the statement of truth, then save and close (during an amend). The next time come back, it'll be "valid" so it'll skip to the pay page even though you never clicked submit (i.e there is no case)

Comment (hosainnet) [reply app/steps/check-your-answers/index.js]: Because in amend we do fast-forward all valid steps..

Comment (dtaylor7) [reply app/steps/check-your-answers/index.js]: ahh, after looking at code i remember now, yes i see what you mean

File: app/core/helpers/steps.js (hunk 6 above)
Comment (dtaylor7) [reply app/core/helpers/steps.js]: infact, may be better to put some code here: https://github.com/hmcts/div-petitioner-frontend/blob/b5c56da3096fd81bc64c85b9859f7cd94159b87a/app/core/helpers/steps.js#L2

would be more performant as wouldn't run through next step logic i.e.
```
const getNextValidStep = function* (step, session) {
  let nextStep; // eslint-disable-line init-declarations
  if (!step.isSkippable) {
    return nextStep;
  }
  ....
```

Comment (hosainnet) [reply app/core/helpers/steps.js]: Correct me if I'm wrong, but I believe that would mean that whatever page we stop at wouldn't have its data pre-populated?

My initial version actually had it a few lines down the lines you've suggested (https://github.com/hmcts/div-petitioner-frontend/pull/497/commits/c619ce0a1b61d09df0b9641c216d7c7eb370d7ad#diff-abc17ad053f837e42773293384c478eaR15) however it turns out `getNextValidStep` is also used within check-your-answers page for some other logic which broke, so I thought `findNextUnAnsweredStep` is a better place because it's the only used in the amend journey (as far as I can tell) and because we _still_ want to carry out the existing validations etc..

Let me know your thoughts, happy for your input as you know this area more :)

Comment (dtaylor7) [reply app/core/helpers/steps.js]: no it only populates data here to do the validation check. but if were always going to stop here then no need to validate.

File: test/end-to-end/codecept.conf.js [hunk 7]
```diff
@@ -75,7 +75,7 @@ function configureChunks() {
 function getTests() {
   console.log('### CONF.preview_env =', CONF.preview_env);  // eslint-disable-line no-console
   if (CONF.preview_env === 'true') {
-    return './paths/**/basicDivorce.js';
+    return './paths/**/*.js';
```
Comment (dtaylor7) [inline test/end-to-end/codecept.conf.js]: also undo this change please

Comment (hosainnet) [reply test/end-to-end/codecept.conf.js]: @vianvi wanted this just to verify the build, will do once all is green!

File: app/core/helpers/steps.js (hunk 6 above)
Comment (hosainnet) [reply app/core/helpers/steps.js]: @dtaylor7 but it'll break this https://github.com/hmcts/div-petitioner-frontend/blob/master/app/steps/check-your-answers/index.js#L208

Comment (dtaylor7) [reply app/core/helpers/steps.js]: yep good point, leave as it is then

File: app/core/steps/Step.js [hunk 8]
```diff
@@ -89,6 +89,10 @@ module.exports = class Step {
     return [true, []];
   }
 
+  isSkipWhenValid() {
```
[the remark this answers is not in the document]
Comment (hosainnet) [reply app/core/steps/Step.js]: Done 👍

File: test/end-to-end/codecept.conf.js (hunk 7 above)
Comment (hosainnet) [reply test/end-to-end/codecept.conf.js]: Done 👍

File: app/core/steps/Step.js [hunk 9]
```diff
@@ -89,6 +89,10 @@ module.exports = class Step {
     return [true, []];
   }
 
+  get isSkippable() {
```
Comment (alexwakeman) [inline app/core/steps/Step.js]: Seems all good, if it was me I would have called this `isRequired` and inverted the logic :)

Comment (dtaylor7) [reply app/core/steps/Step.js]: good point, always more readable when testing for truthy bools
